In [1]:
import numpy as np, pandas as pd, seaborn as sns

df = sns.load_dataset('titanic')
df.to_csv('../data/titanic.csv', index=False)

Возьми salaries = np.array([30, 32, 35, 35, 38, 40, 42]) — зарплаты в тысячах. Посчитай среднее без np.mean — через сумму и количество. И медиану без np.median — через сортировку и индекс. Сверь оба с библиотечными функциями.

In [2]:
salaries = np.array([30, 32, 35, 35, 38, 40])
def my_mean(vector):
    total = 0
    for i in vector:
        total += i
    return total / len(vector)

def my_median(vector):
    sorted_vect = sorted(vector)
    if len(vector) % 2 != 0:
        return sorted_vect[len(vector) // 2]
    else:
        return ((sorted_vect[(len(vector) // 2) - 1]) + (sorted_vect[len(vector) // 2])) / 2

print(salaries.mean(), my_mean(salaries))
print(np.median(salaries), my_median(salaries))

35.0 35.0
35.0 35.0


Добавь в этот массив одну зарплату директора: 500. Посчитай среднее и медиану заново. На сколько сдвинулась каждая? Сформулируй словами, какая устойчива к выбросу и почему — через определения выше, а не «потому что так работает»

In [3]:
salaries = np.append(salaries, 500)
print(my_mean(salaries), my_median(salaries))

101.42857142857143 35


По тесту видно что медиана не сильно сдвинулась в отличии от среднего значения. Медиана является более устойчива к выбросу так как при расчете не смотрит на значения, а тольеко на индексы

Проверь на fare из Titanic: среднее, медиана. Разрыв объясни. Какую величину ты бы поставил в отчёт с подписью «типичная цена билета» и почему?

In [4]:
print(my_mean(df['fare']), my_median(df['fare']))

32.2042079685746 14.4542


Разрыв медианы и среднего почти в два раза. Связанно с тем что есть люди цена билетов слишком велика, но такие люди редкость, медиана показывает цену билета который стоит в середине топа билетов по ценам так сказать поэтому я бы его поставил в отчёт, так как он более точно отражает типичную цену билета

Медиана массива из чётного числа элементов — что это за значение? Проверь на [1, 2, 3, 4]: посчитай np.median и подумай, откуда взялся ответ, если такого элемента в массиве нет.

In [5]:
np.median([1, 2, 3, 4])

np.float64(2.5)

Ответ 2.5. Потому список состоит из четного количества эллементов, отсюда следует, что по середине стоит два эллемента, при поиске медианы функция нашла среднее между этими числами

Посчитай моду для колонки embarked и для колонки fare. Сравни, насколько осмысленны оба ответа. Для fare посмотри заодно, сколько раз встречается самое частое значение и сколько всего разных значений в колонке

In [6]:
print(df['embarked'].mode(), df['fare'].mode(), sep='\n')

0    S
Name: embarked, dtype: str
0    8.05
Name: fare, dtype: float64


In [7]:
df[df['fare'] == 8.05]

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
4,0,3,male,35.0,0,0,8.05,S,Third,man,True,NaN,Southampton,no,True
12,0,3,male,20.0,0,0,8.05,S,Third,man,True,NaN,Southampton,no,True
37,0,3,male,21.0,0,0,8.05,S,Third,man,True,NaN,Southampton,no,True
45,0,3,male,NaN,0,0,8.05,S,Third,man,True,NaN,Southampton,no,True
77,0,3,male,NaN,0,0,8.05,S,Third,man,True,NaN,Southampton,no,True
87,0,3,male,NaN,0,0,8.05,S,Third,man,True,NaN,Southampton,no,True
89,0,3,male,24.0,0,0,8.05,S,Third,man,True,NaN,Southampton,no,True
90,0,3,male,29.0,0,0,8.05,S,Third,man,True,NaN,Southampton,no,True
95,0,3,male,NaN,0,0,8.05,S,Third,man,True,NaN,Southampton,no,True
112,0,3,male,22.0,0,0,8.05,S,Third,man,True,NaN,Southampton,no,True


In [8]:
df['fare'].nunique()

248

Мода для embarked осмылсенна ведь там всего 3 значения не считая nan, а вот для fare ведь там 248 разных значений и в итоге модой оказалось значение которое встречется 43 раза из 891

Попробую посчитать среднее отклонение от среднего значения для массива,
Ожидаю результат ноль

In [9]:
B = np.array([5, 35, 65])
def f1(B):
    mean_b = B.mean()
    res = 0
    for x in B:
        res += x - mean_b
    return res
print(f1(B))

0.0


Получился ноль ведь две половины компенсировали друг друга
Проверю на других массивах

In [10]:
print(f1(np.array([12, 32, 324, 321, 1, 435, 3])))
print(f1(np.array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1])))
print(f1(df['fare']))

5.684341886080802e-14
0.0
-2.3092638912203256e-13


Почему то в 1 и 3 получились числа очень близкие к нулю скорее всего дело в особенности массивов

Напишу подсчет среднего отклонения через модуль и через квадрат

In [11]:
def my_abs_std(vector):
    mean_v = vector.mean()
    res = 0
    for x in vector:
        res += abs(x - mean_v)
    return res / len(vector)

def my_std(vector):
    mean_v = vector.mean()
    res = 0
    for x in vector:
        res += (x - mean_v)**2
    return (res / len(vector))**(0.5)

print(my_abs_std(B))
print(my_std(B))
B[2] = 200
print(my_abs_std(B))
print(my_std(B))

20.0
24.49489742783178
80.0
85.73214099741124


На более далекие отклонения реагирует сильнее способ при помощи модулю, результат отличается в 4 раза
При помощи квадрата чуть меньше чем в 4 раза

In [12]:
print(np.std(salaries), my_std(salaries))
print(np.std(df['fare']), my_std(df['fare']))

162.7459520171214 162.7459520171214
49.6655344447741 49.6655344447741


Результат моей функции не отличается от результата функции из NumPy

In [14]:
x = np.array([2., 4, 4, 4, 5, 5, 7, 9])
print(np.std(x), pd.Series(x).std())

2.0 2.138089935299395


Я предсказывал одинаковый результат но получился разный, видимо нумпаевский способ посчитал с погрешностью

In [15]:
print(my_std(x))

2.0


Оказывается результат нумпая верен, но тогда не понятно почему результат 2 неверен
Попробую делить в своей функции не на n, а на n - 1

In [16]:
def my_std2(vector):
    mean_v = vector.mean()
    res = 0
    for x in vector:
        res += (x - mean_v)**2
    return (res / (len(vector) - 1))**(0.5)
print(my_std2(x))

2.138089935299395


Тогда понятно, результат нумпая получается путем деления суммы на n когда во 2 случае сумма делится на n - 1

In [30]:
rng = np.random.default_rng()
a = rng.normal(100, 15, 1_000_000)
print(np.std(a))
one = np.array([])
two = np.array([])
for _ in range(20000):
    choice = np.random.choice(a, size=30, replace=True)
    one = np.append(one, np.std(choice, ddof=0))
    two = np.append(two, np.std(choice, ddof=1))

print(one.mean(), two.mean())

15.00339216910481
14.632142598751884 14.88228281398352


2 оценка оказалась ближе к истине, когда первая оказалась чуть дальше в минус, при работе стоит смотреть на оценки с ddof=1, при увелечении выборки растет точность и падает разница между оценками